File: iterated-popcount-results-in-0-or-1/iteratedpopcount.py
# 
# Iterated popcount demo
# 
# Copyright (c) 2014 Nayuki Minase
# All rights reserved. Contact Nayuki for licensing.
# http://nayuki.eigenstate.org/page/iterated-popcount-results-in-0-or-1
# 

import sys
if sys.version_info.major == 2:
    input = raw_input


def main():
    while True:
        s = input("Enter an integer (or blank to quit): ")
        if s == "":
            print("Quit")
            break
        try:
            n = int(s)
            if n < 0:
                raise ValueError()
            do_iterated_popcount(n)
        except ValueError:
            print("Error: Number must be positive or zero")
        print("")


def do_iterated_popcount(n):
    i = 0
    prev = -1
    print("Iter  Value")
    while True:
        print("{:4}  {}".format(i, n))
        if n == prev:
            break
        prev = n
        n = bin(n).count("1")
        i += 1


if __name__ == "__main__":
    main()

File: iterated-popcount-results-in-0-or-1/IteratedPopcount.java
/* 
 * Iterated popcount demo
 * 
 * Copyright (c) 2017 Project Nayuki
 * All rights reserved. Contact Nayuki for licensing.
 * https://www.nayuki.io/page/iterated-popcount-results-in-0-or-1
 */

import java.math.BigInteger;
import java.util.Scanner;


public final class IteratedPopcount {
	
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		while (true) {
			System.out.print("Enter an integer (or blank to quit): ");
			System.out.flush();
			String line = scan.nextLine();
			if (line.equals("")) {
				System.out.println("Quit");
				break;
			}
			try {
				BigInteger n = new BigInteger(line);
				if (n.signum() == -1)
					throw new NumberFormatException();
				doIteratedPopcount(n);
			} catch (NumberFormatException e) {
				System.out.println("Error: Number must be positive or zero");
			}
			System.out.println();
		}
	}
	
	
	private static void doIteratedPopcount(BigInteger n) {
		int i = 0;
		BigInteger prev = BigInteger.valueOf(-1);
		System.out.println("Iter  Value");
		while (true) {
			System.out.printf("%4d  %d%n", i, n);
			if (n.equals(prev))
				break;
			prev = n;
			n = BigInteger.valueOf(n.bitCount());
			i++;
		}
	}
	
}
